### PR TITLE
Remove draft articles with duplicate feed source URL and cleanup invalid data: take 2

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

**this PR is a clone of https://github.com/forem/forem/pull/10205 that we ended up reverting as it was merged prematurely**

Similar to https://github.com/forem/forem/pull/10138 but based on `feed_source_url`.

To be able to close the loop on #8072 and add a unique index on feed_source_url we need first to be remove duplicates that have the same feed_source_url.

This PR contains scripts to accomplish the following things:

1. remove all drafts that in a partition have the same body. A partition is determined as a group of articles with the same feed source URL
2. remove all drafts that in a partition have different bodies by keeping the most recent one. A partition is determined as a group of articles with the same feed source URL
3. fixes invalid data for `feed_source_url`. Currently, unlike with `canonical_url`, we don't have a URL validator on the field, thus I found a bunch of invalid URLs in production data, see https://dev.to/admin/blazer/queries/220-articles-with-an-invalid-feed-source-url

Please read the SQL queries closely.

If you want to take a look at the data:

- https://dev.to/admin/blazer/queries/218-duplicate-articles-feed_source_url-with-the-same-body
- https://dev.to/admin/blazer/queries/219-duplicate-articles-feed_source_url-with-different-bodies
- https://dev.to/admin/blazer/queries/220-articles-with-an-invalid-feed-source-url

## Related Tickets & Documents

#10138 
#8072 

## QA Instructions, Screenshots, Recordings

Ideally this should be tested on a copy of the production DB. To make sure it worked I downloaded the sample of duplicates https://dev.to/admin/blazer/queries/221-duplicate-articles-feed_source-url-with-the-same-body-ids - imported them locally and ran the scripts to make sure they worked. They obviously present different results as it's a script run on the subset of duplications (so it works because in some cases we have 2 or more duplicates of the same article).

If you want to test it nonetheless:

- download a CSV from https://dev.to/admin/blazer/queries/221-duplicate-articles-feed_source-url-with-the-same-body-ids
- open the SQL console with `psql PracticalDeveloper_development`
- type `\d articles`
- note the names of the foreign key constraints
- drop the foreign key constraints on the `articles` table, otherwise the `user_id` and `organization_id` won't work:

```sql
alter table articles drop constraint fk_rails_3d31dad1cc;
alter table articles drop constraint fk_rails_7809a1a57d;
```

remember that the constraint names might be different than mine. Please double check with `\d articles` in the SQL console and replace them with yours
- import the data from the downloaded CSV:

```sql
copy articles(id,user_id,title,body_html,slug,created_at,updated_at,main_image,description,published,published_at,featured,processed_html,featured_number,social_image,body_markdown,organization_id,canonical_url,collection_id,show_comments,main_image_background_hex_color,receive_notifications,approved,published_from_feed,comments_count,reactions_count,video,video_code,video_source_url,video_thumbnail_url,video_closed_caption_track_url,hotness_score,feed_source_url,second_user_id,third_user_id,last_buffered,positive_reactions_count,edited_at,crossposted_at,spaminess_rating,language,facebook_last_buffered,cached_tag_list,path,cached_user_name,cached_user_username,last_comment_at,boost_states,video_state,page_views_count,previous_positive_reactions_count,score,reading_time,comment_template,video_duration_in_seconds,experience_level_rating,experience_level_rating_distribution,last_experience_level_rating_at,rating_votes_count,originally_published_at,organic_page_views_count,organic_page_views_past_month_count,organic_page_views_past_week_count,cached_user,cached_organization,archived,any_comments_hidden,nth_published_by_author,comment_score,search_optimized_title_preamble,search_optimized_description_replacement,public_reactions_count,previous_public_reactions_count,user_subscriptions_count) from '/path/to/duplicate-articles-feed_source-url-with-the-same-body-ids.csv' DELIMITER ',' csv header;
```

don't forget to replace the path of the file with the full path of the file you downloaded

- count the number of articles, just to have a quick idea if the script has worked
- run the data update script with `rails data_updates:run` if it's the first time or manually in the console

```ruby
require Rails.root.join("lib/data_update_scripts/20200904132553_remove_draft_articles_with_duplicate_feed_source_url.rb")
DataUpdateScripts::RemoveDraftArticlesWithDuplicateFeedSourceUrl.new.run
```

- re-count the number of articles, it should be less ;-)
- run https://dev.to/admin/blazer/queries/218-duplicate-articles-feed_source_url-with-the-same-body  locally to make sure they are gone

If you would like to test the cleanup script as well, you can do something similar by using:

- https://dev.to/admin/blazer/queries/222-articles-with-an-invalid-feed-source-url-select as the source data
- the following to test:

```ruby
require Rails.root.join("lib/data_update_scripts/20200904141057_cleanup_articles_with_invalid_feed_source_url.rb")
DataUpdateScripts::CleanupArticlesWithInvalidFeedSourceUrl.new.run
```

- re-run https://dev.to/admin/blazer/queries/220-articles-with-an-invalid-feed-source-url locally to check the data has been cleaned

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

### CODE REVIEW

**Comment by mstruve** [comment]
Forking production database to test these queries

**Review by mstruve — APPROVED** [review]
Fork results match our production data!
result_same_body deleted 4469 records
result_different_bodies deleted 3 records

Double-checked with select statements that they both return 0 now. Lets do this! 🚀

File: lib/data_update_scripts/20200904141057_cleanup_articles_with_invalid_feed_source_url.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module DataUpdateScripts
+  class CleanupArticlesWithInvalidFeedSourceUrl
+    def run
+      # We need this to intercept those strings that would be valid URLs if the scheme were added,
+      # so I slightly modified the Ruby URL regexp from https://urlregex.com/ to avoid checking for the scheme
+      almost_url_regexp = %r{\A(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:/[^\s]*)?\z}i # rubocop:disable Layout/LineLength
```
Comment (citizen428) [inline lib/data_update_scripts/20200904141057_cleanup_articles_with_invalid_feed_source_url.rb:6]: You can use the `x` modifier on this regex to break it over multiple lines if you want, similar to how `URI.regex` does it:

```ruby
/
        ([a-zA-Z][\-+.a-zA-Z\d]*):                           (?# 1: scheme)
        (?:
           ((?:[\-_.!~*'()a-zA-Z\d;?:@&=+$,]|%[a-fA-F\d]{2})(?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*)                    (?# 2: opaque)
        |
           (?:(?:
             \/\/(?:
                 (?:(?:((?:[\-_.!~*'()a-zA-Z\d;:&=+$,]|%[a-fA-F\d]{2})*)@)?        (?# 3: userinfo)
                   (?:((?:(?:[a-zA-Z0-9\-.]|%\h\h)+|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|\[(?:(?:[a-fA-F\d]{1,4}:)*(?:[a-fA-F\d]{1,4}|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|(?:(?:[a-fA-F\d]{1,4}:)*[a-fA-F\d]{1,4})?::(?:(?:[a-fA-F\d]{1,4}:)*(?:[a-fA-F\d]{1,4}|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}))?)\]))(?::(\d*))?))? (?# 4: host, 5: port)
               |
                 ((?:[\-_.!~*'()a-zA-Z\d$,;:@&=+]|%[a-fA-F\d]{2})+)                 (?# 6: registry)
               )
             |
             (?!\/\/))                           (?# XXX: '\/\/' is the mark for hostport)
             (\/(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*(?:;(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*)*(?:\/(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*(?:;(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*)*)*)?                    (?# 7: path)
           )(?:\?((?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*))?                 (?# 8: query)
        )
        (?:\#((?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*))?                  (?# 9: fragment)
      /x
```

**Review by citizen428 — APPROVED** [review]

File: lib/data_update_scripts/20200904141057_cleanup_articles_with_invalid_feed_source_url.rb (hunk 1 above)
Comment (rhymes) [reply lib/data_update_scripts/20200904141057_cleanup_articles_with_invalid_feed_source_url.rb:6]: ahhaha not sure if it's worth it, it's still quite hard to read. If it were in the app itself I would probably use a different strategy but for one time cleanup it's probably okay to keep it as is with the inline comment

**Pull request merged**